src/backend/marsha/core/tests/test_models_live_pairing.py

```diff
@@ -0,0 +1,35 @@
+"""Tests for the models in the ``core`` app of the Marsha project."""
+from django.test import TestCase
+
+from ..factories import LivePairingFactory
+from ..models import LivePairing
+
+
+class LivePairingModelsTestCase(TestCase):
+    """Test our intentions about the LivePairing model."""
+
+    def test_models_live_pairing_str(self):
+        """The str method should display the title of the video."""
+        live_pairing = LivePairingFactory(video__title="j'espère")
+        self.assertEqual(str(live_pairing), "j'espère")
+
+    def test_models_live_pairing_deleted(self):
+        """Deleting a LivePairing should be hard."""
+        live_pairing = LivePairingFactory()
+
+        live_pairing.delete()
+        with self.assertRaises(LivePairing.DoesNotExist):
+            live_pairing.refresh_from_db()
+
+    def test_models_live_pairing_secret_generator(self):
+        """The secret_generator should return a 6 digit string.
+
+        The secret should not exist in existing_secrets parameter."""
+        # Generate a list of all 6 digits strings
+        existing_secrets = list(str(i).zfill(6) for i in range(0, 1000000))
+        # Remove one from the list
+        existing_secrets.remove("123456")
+
+        # Ensure removed secret has been found and used
```

Comment:
Actually this test is scary because it demonstrates a case where the secret is directly guessable 😅 as mentioned above we need to protect against this situation.

Reply:
I don't get how the secret is guessable…

Reply:
Because if all possible secrets are existing, then whatever secret I submit, it will match some video... In order for this pairing to be secure, we must limit the number of existing secrets at a given time... maybe 10 or 100? And you shouldn't return a different error if the secret has expired or is unknown or has throttled... always return a 404 is safer.

Reply:
I thing the try/except I proposed instead of looking for the existing secrets, is solving this issue because the try/except will fail when the number of secrets is too much...

Reply:
What about creating a management command deleting all secret expired every hour ? Using redis we can use an expire key 😁

Reply:
I don't think `secret_generator` should be responsible for security aspects. His job is only to return an available secret.
And, I agree, if 999 999 secrets are existing, we have a problem which should be addressed, but not here.
This test reflects only the performance of my implementation.

Reply:
Expired secrets do not pause any problem if you don't leak the information on the error and if you add a conditional unique constraint that only checks unicity on valid secrets?

Reply:
secrets are already unique, see LivePairing model :
```
    secret = models.CharField(
        max_length=6,
        unique=True,
        verbose_name=_("secret"),
        help_text=_("live pairing secret"),
    )
```
Maybe we could  simply delete all expired secrets on every request ? I think it should be fast enough, and not noticed by the users.
But I agree with you, we shouldn't give any clues on errors.

Reply:
Yes good idea, start by deleting all expired secrets and let's avoid another cron :sweat_smile:

Reply:
Are we ok with the code related to this thread ?

Reply:
OK for the delete but I still think the code retrieving existing secrets does not work with concurrent requests and should be based on a try/except pattern.

Reply:
Should we test concurrent requests then ?

Reply:
well... that's hard... but the try/except pattern surely solves it.